File: trunk/src/test/org/codehaus/testdox/intellij/panel/ItemSelectionDialogTest.java
package org.codehaus.testdox.intellij.panel;

import com.intellij.openapi.project.Project;
import org.intellij.openapi.testing.DialogCreator;
import org.intellij.openapi.testing.MockApplicationManager;
import org.jmock.MockObjectTestCase;

import java.awt.*;

public class ItemSelectionDialogTest extends MockObjectTestCase {

    protected void setUp() {
        MockApplicationManager.clear();
    }

    public void testEnablesOkButtonIfSelectionIsMade() throws Exception {
        ItemSelectionDialog dialog = createDialog("hi", "bye");
        dialog.getList().addSelectionInterval(0, 0);
        assertTrue(dialog.isOKActionEnabled());
    }

    public void testInitiallyDisablesOkButton() throws Exception {
        assertFalse(createDialog("hi", "bye").isOKActionEnabled());
    }

    public void testReturnsCancelledWithoutShowingItselfIfNoItemsWereProvided() throws Exception {
        ItemSelectionDialog dialog = createDialog();
        show(dialog);
        assertEquals(ItemSelectionDialog.CANCEL_EXIT_CODE, dialog.getExitCode());
    }

    public void testReturnsOkWithoutShowingDialogIfOnlyOneItemWasProvided() throws Exception {
        String item = "hi";
        ItemSelectionDialog dialog = createDialog(item);
        show(dialog);
        assertEquals(true, dialog.isOK());
        assertEquals(item, dialog.getSelectedItem());
    }

    private ItemSelectionDialog createDialog(final Object... items) throws Exception {
        DialogCreator<ItemSelectionDialog> dialogCreator = new DialogCreator<ItemSelectionDialog>() {
            protected ItemSelectionDialog create() {
                return new ItemSelectionDialog((Project) mock(Project.class).proxy(), items, "blah", "blah", null);
            }
        };
        EventQueue.invokeAndWait(dialogCreator);
        return dialogCreator.getDialog();
    }

    private void show(final ItemSelectionDialog dialog) throws Exception {
        EventQueue.invokeAndWait(new Runnable() {
            public void run() {
                dialog.show();
            }
        });
    }
}

File: test/scala/org/codehaus/testdox/intellij/ui/ItemSelectionDialogTest.scala
package org.codehaus.testdox.intellij.ui

import com.intellij.openapi.project.Project
import org.intellij.openapi.testing.DialogCreator
import org.intellij.openapi.testing.MockApplicationManager
import org.jmock.MockObjectTestCase

import java.awt._
import junit.framework.Assert._
import com.intellij.openapi.ui.DialogWrapper.CANCEL_EXIT_CODE

class ItemSelectionDialogTest extends MockObjectTestCase {

  override protected def setUp() {
    MockApplicationManager.clear()
  }

  def testEnablesOkButtonIfSelectionIsMade() {
    val dialog = createDialog("hi", "bye")
    dialog.getList().addSelectionInterval(0, 0)
    assertTrue(dialog.isOKActionEnabled())
  }

  def testInitiallyDisablesOkButton() {
    assertFalse(createDialog("hi", "bye").isOKActionEnabled())
  }

  def testReturnsCancelledWithoutShowingItselfIfNoItemsWereProvided() {
    val dialog = createDialog()
    show(dialog)
    assertEquals(CANCEL_EXIT_CODE, dialog.getExitCode())
  }

  def testReturnsOkWithoutShowingDialogIfOnlyOneItemWasProvided() {
    val item = "hi"
    val dialog = createDialog(item)
    show(dialog)
    assertEquals(true, dialog.isOK())
    assertEquals(item, dialog.getSelectedItem())
  }

  private def createDialog(items: Object*): ItemSelectionDialog = {
    val dialogCreator = new DialogCreator[ItemSelectionDialog]() {
      def create() = new ItemSelectionDialog(mock(classOf[Project]).proxy().asInstanceOf[Project], items.toArray[Object], "blah", "blah", null)
    }
    EventQueue.invokeAndWait(dialogCreator)
    dialogCreator.getDialog()
  }

  private def show(dialog: ItemSelectionDialog) {
    EventQueue.invokeAndWait(new Runnable() {
      def run() = dialog.show()
    })
  }
}
